cpp/src/solver/lars_impl.cuh

```diff
@@ -0,0 +1,705 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <iostream>
+#include <limits>
+#include <numeric>
+
+#include <raft/cudart_utils.h>
+#include <raft/linalg/cublas_wrappers.h>
+#include <raft/linalg/gemv.h>
+#include <thrust/copy.h>
+#include <thrust/device_ptr.h>
+#include <thrust/execution_policy.h>
+#include <thrust/iterator/counting_iterator.h>
+#include <thrust/sequence.h>
+#include <thrust/sort.h>
+#include <cache/cache_util.cuh>
+#include <common/allocatorAdapter.hpp>
+#include <common/cumlHandle.hpp>
+#include <common/device_buffer.hpp>
+#include <common/host_buffer.hpp>
+#include <cub/cub.cuh>
+#include <cuml/common/logger.hpp>
+#include <raft/cuda_utils.cuh>
+#include <raft/linalg/binary_op.cuh>
+#include <raft/linalg/cholesky_r1_update.cuh>
+#include <raft/linalg/map_then_reduce.cuh>
+#include <raft/linalg/unary_op.cuh>
+
+namespace ML {
+namespace Solver {
+namespace Lars {
+
+/**
+ * @brief Select the largest element from the inactive working set.
+ *
+ * The inactive set consist of cor[n_active..n-1]. This function returns the
+ * index of the most correlated element. The value of the largest element is
+ * returned in cj.
+ *
+ * @param n_active number of active elements (n_active <= n )
+ * @param n number of elements in vector cor
+ * @param correlation device array of correlations, size [n]
+ * @param cj host pointer to return the value of the largest element
+ * @param stream CUDA stream
+ */
+template <typename math_t, typename idx_t = int>
+idx_t selectMostCorrelated(idx_t n_active, idx_t n, math_t* correlation,
+                           math_t* cj,
+                           MLCommon::device_buffer<math_t>& workspace,
+                           cudaStream_t stream) {
+  const idx_t align_bytes = 16 * sizeof(math_t);
+  // We might need to start a few elements earlier to ensure that the unary
```

Comment:
we discussed this during our 1:1 today and unary-op does look at the proper alignment of the input addresses. So, this extra alignment shouldn't be needed, right?